File: haskell/src/Surd/Galois/TransitiveGroup.hs
{- |
Module      : Surd.Galois.TransitiveGroup
Description : Database of transitive subgroups of \(S_n\) for small \(n\)
Stability   : experimental
License     : BSD-3-Clause

= Background

A /transitive subgroup/ of \(S_n\) is a permutation group acting on
\(\{1, \ldots, n\}\) whose action has a single orbit.  The transitive
subgroups of \(S_n\) (up to conjugacy) classify the possible Galois groups
of irreducible degree-\(n\) polynomials over \(\mathbb{Q}\).

For Galois group identification via Stauduhar's resolvent method, we need
the lattice of transitive subgroups together with generators, orders,
solvability flags, and maximal-supergroup relationships.

== Numbering convention

Groups are numbered following the Butler–McKay convention (1983), which
is also the standard in GAP, Magma, and the LMFDB.  Within a given
degree, \(T_1\) is the smallest group and \(T_k\) (where \(k\) is the
total number of conjugacy classes) is \(S_n\) itself.

== Degree-5 transitive group lattice

@
          S5  (T5, order 120)
         \/  \\
        A5    \\   (T4, order 60)
              F20  (T3, order 20, solvable)
               |
              D5   (T2, order 10, solvable)
               |
              C5   (T1, order  5, solvable)
@

Lines denote containment (subgroup below, supergroup above).
\(F_{20} = \mathbb{Z}/5 \rtimes \mathbb{Z}/4\) is the Frobenius group,
the largest /solvable/ transitive subgroup of \(S_5\).

== Solvability and radical solvability

A finite group \(G\) is /solvable/ if it admits a composition series

\[
  G = G_0 \trianglerighteq G_1 \trianglerighteq \cdots \trianglerighteq G_k = \{1\}
\]

where each quotient \(G_i / G_{i+1}\) is cyclic of prime order.
By Galois's theorem, a polynomial is solvable by radicals if and only if
its Galois group is solvable.  The composition factors determine the
radical tower: a cyclic quotient of order \(p\) corresponds to adjoining
a \(p\)-th root.  For degree 5, the groups \(C_5\), \(D_5\), and
\(F_{20}\) are solvable, while \(A_5\) and \(S_5\) are not.

== References

* Butler, G. & McKay, J. (1983). "The transitive groups of degree up to
  eleven." /Communications in Algebra/ 11(8), 863–911.
  DOI: 10.1080\/00927878308822884

* Hulpke, A. (2005). "Constructing transitive permutation groups."
  /J. Symbolic Computation/ 39(1), 1–30.
  DOI: 10.1016\/j.jsc.2004.08.002

* Cox, D. A. (2012). /Galois Theory/, 2nd ed. Wiley.
  DOI: 10.1002\/9781118218457
-}
module Surd.Galois.TransitiveGroup (
    -- * Transitive group type
    TransitiveGroup (..),

    -- * Database queries
    transGroupsOfDegree,
    transGroupByOrder,

    -- * Solvability
    isSolvable,

    -- * Composition series
    compositionSeries,
)
where

import Data.List (nub, sort)
import Math.Internal.Positive (Positive)
import Math.Internal.PrimeFactors (factorise)
import Surd.Galois.Permutation

{- | A transitive subgroup of \(S_n\), with metadata for Galois group
identification and radical tower construction.
-}
data TransitiveGroup = TransitiveGroup
    { tgName :: !String
    -- ^ Human-readable name, e.g. @\"C5\"@, @\"D5\"@, @\"F20\"@, @\"A5\"@, @\"S5\"@.
    , tgDegree :: !Int
    {- ^ The degree \(n\) of the symmetric group \(S_n\) in which this
    group acts (equivalently, the number of roots of the polynomial).
    -}
    , tgOrder :: !Integer
    -- ^ The order \(|G|\) of the group.
    , tgGenerators :: ![Perm]
    {- ^ A generating set for the group, given as permutations in cycle
    notation on the symbols \(\{0, 1, \ldots, n-1\}\).  For example,
    \(D_5 = \langle (0\;1\;2\;3\;4),\; (1\;4)(2\;3) \rangle\).
    -}
    , tgSolvable :: !Bool
    {- ^ Whether the group is solvable.  A group is solvable iff it has a
    composition series with cyclic (prime order) quotients, which by
    Galois's theorem is equivalent to the corresponding polynomial being
    solvable by radicals.
    -}
    , tgMaximalSupergroups :: ![Int]
    {- ^ Indices (0-based) into the list returned by 'transGroupsOfDegree'
    of groups that contain this group as a maximal transitive subgroup.
    This encodes the Hasse diagram of the transitive group lattice and
    drives the Stauduhar descent strategy: begin with \(S_n\) (the top
    of the lattice) and descend via resolvent polynomials, testing each
    maximal subgroup until the Galois group is identified.  Empty for
    \(S_n\) itself (the top element).
    -}
    , tgCompositionFactors :: ![Int]
    {- ^ For solvable groups: the prime orders of the cyclic quotients in a
    composition series

    \[
    G = G_0 \trianglerighteq G_1 \trianglerighteq \cdots
      \trianglerighteq G_k = \{1\}
    \]

    listed from top to bottom.  For example, @[5, 2, 2]@ for \(F_{20}\)
    means \(F_{20} / D_5 \cong \mathbb{Z}/2\),
    \(D_5 / C_5 \cong \mathbb{Z}/2\), \(C_5 / \{1\} \cong \mathbb{Z}/5\)
    (reading the factors in reverse gives the radical tower from innermost
    to outermost root extraction).  Empty for non-solvable groups.
    -}
    }
    deriving (Show, Eq)

{- | Test whether a transitive group is solvable (\(\equiv\) 'tgSolvable').
A solvable Galois group means the polynomial is solvable by radicals,
i.e.\ its roots can be expressed using nested \(p\)-th roots over
\(\mathbb{Q}\).
-}
isSolvable :: TransitiveGroup -> Bool
isSolvable = tgSolvable

{- | All transitive subgroups of \(S_n\) (up to conjugacy) for the given
degree \(n\), sorted by increasing group order.

For prime \(n\), the groups are computed at runtime from the structure
of \(\mathrm{AGL}(1,n)\). For degree 5, the hard-coded database is
used (as a fast path). For unsupported composite degrees, returns the
empty list.
-}
transGroupsOfDegree :: Int -> [TransitiveGroup]
transGroupsOfDegree 5 = degree5Groups
transGroupsOfDegree n
    | n >= 3 && isPrime (fromIntegral n) = transGroupsOfPrimeRT n
    | otherwise = []

-- | Test whether an integer is prime (trial division, sufficient for
-- degrees we care about).
isPrime :: Integer -> Bool
isPrime n
    | n < 2 = False
    | n < 4 = True
    | even n = False
    | otherwise = all (\d -> n `mod` d /= 0) [3, 5 .. isqrt n]
  where
    isqrt = floor . (sqrt :: Double -> Double) . fromIntegral

{- | Find the transitive group(s) of the given degree and order.
There may be more than one conjugacy class with the same order
(though this does not occur in degree 5).
-}
transGroupByOrder :: Int -> Integer -> [TransitiveGroup]
transGroupByOrder deg ord =
    [g | g <- transGroupsOfDegree deg, tgOrder g == ord]

------------------------------------------------------------------------
-- Degree 5 transitive groups
------------------------------------------------------------------------

-- The five transitive subgroups of S_5, in order of increasing order:
--
-- T1 = C5  (Z/5Z)       order  5   solvable
-- T2 = D5  (Dih(5))     order 10   solvable
-- T3 = F20 (Z/5 ⋊ Z/4) order 20   solvable
-- T4 = A5  (Alt(5))     order 60   not solvable
-- T5 = S5  (Sym(5))     order 120  not solvable
--
-- Labelling of roots: {0, 1, 2, 3, 4}.
-- Generators use the standard (0-indexed) permutation representation.

degree5Groups :: [TransitiveGroup]
degree5Groups =
    [ -- T1: C5 = ⟨(0 1 2 3 4)⟩
      TransitiveGroup
        { tgName = "C5"
        , tgDegree = 5
        , tgOrder = 5
        , tgGenerators = [fromCycles 5 [[0, 1, 2, 3, 4]]]
        , tgSolvable = True
        , tgMaximalSupergroups = [1] -- D5
        , tgCompositionFactors = [5]
        }
    , -- T2: D5 = ⟨(0 1 2 3 4), (1 4)(2 3)⟩
      TransitiveGroup
        { tgName = "D5"
        , tgDegree = 5
        , tgOrder = 10
        , tgGenerators =
            [ fromCycles 5 [[0, 1, 2, 3, 4]]
            , fromCycles 5 [[1, 4], [2, 3]]
            ]
        , tgSolvable = True
        , tgMaximalSupergroups = [2] -- F20
        , tgCompositionFactors = [5, 2]
        }
    , -- T3: F20 = GA(1,5) = ⟨(0 1 2 3 4), (1 2 4 3)⟩
      -- The Frobenius group of order 20 = Z/5 ⋊ Z/4.
      -- (1 2 4 3) acts as multiplication by 2 mod 5 on {0,1,2,3,4}.
      TransitiveGroup
        { tgName = "F20"
        , tgDegree = 5
        , tgOrder = 20
        , tgGenerators =
            [ fromCycles 5 [[0, 1, 2, 3, 4]]
            , fromCycles 5 [[1, 2, 4, 3]]
            ]
        , tgSolvable = True
        , tgMaximalSupergroups = [3, 4] -- A5, S5
        , tgCompositionFactors = [5, 2, 2]
        }
    , -- T4: A5 = ⟨(0 1 2 3 4), (0 1 2)⟩
      TransitiveGroup
        { tgName = "A5"
        , tgDegree = 5
        , tgOrder = 60
        , tgGenerators =
            [ fromCycles 5 [[0, 1, 2, 3, 4]]
            , fromCycles 5 [[0, 1, 2]]
            ]
        , tgSolvable = False
        , tgMaximalSupergroups = [4] -- S5
        , tgCompositionFactors = []
        }
    , -- T5: S5 = ⟨(0 1 2 3 4), (0 1)⟩
      TransitiveGroup
        { tgName = "S5"
        , tgDegree = 5
        , tgOrder = 120
        , tgGenerators =
            [ fromCycles 5 [[0, 1, 2, 3, 4]]
            , fromCycles 5 [[0, 1]]
            ]
        , tgSolvable = False
        , tgMaximalSupergroups = [] -- top
        , tgCompositionFactors = []
        }
    ]

------------------------------------------------------------------------
-- Runtime prime group computation
------------------------------------------------------------------------

-- | Modular exponentiation (internal).
modExpRT :: Integer -> Integer -> Integer -> Integer
modExpRT _ 0 _ = 1
modExpRT b e m
    | even e = let half = modExpRT b (e `div` 2) m in (half * half) `mod` m
    | otherwise = (b * modExpRT b (e - 1) m) `mod` m

-- | Primitive root modulo a prime (internal).
primitiveRootRT :: Integer -> Integer
primitiveRootRT p =
    let phi = p - 1
        factors = nub $ map fst (factorise (fromInteger phi :: Positive))
        isPrimRoot g = all (\q -> modExpRT g (phi `div` q) p /= 1) factors
     in head (filter isPrimRoot [2 .. p - 1])

-- | Sorted positive divisors of n (internal).
divisorsRT :: Integer -> [Integer]
divisorsRT n =
    let isqrt = floor . (sqrt :: Double -> Double) . fromIntegral
        small = [d | d <- [1 .. isqrt n], n `mod` d == 0]
     in nub $ sort $ concatMap (\d -> if d * d == n then [d] else [d, n `div` d]) small

-- | Prime factorisation as a flat list with multiplicity (internal).
primeFactorsWithMultRT :: Integer -> [Integer]
primeFactorsWithMultRT 1 = []
primeFactorsWithMultRT n =
    concatMap (\(q, e) -> replicate e q) (factorise (fromInteger n :: Positive))

{- | All transitive subgroups of \(S_p\) for a prime \(p\), computed at
runtime from the structure of \(\mathrm{AGL}(1,p)\).

The solvable subgroups are \(\mathbb{Z}/p \rtimes H\) for each divisor
\(d\) of \(p-1\), plus \(A_p\) and \(S_p\) (non-solvable for \(p \ge 5\)).
-}
transGroupsOfPrimeRT :: Int -> [TransitiveGroup]
transGroupsOfPrimeRT p' =
    let p = fromIntegral p' :: Integer
        g = primitiveRootRT p
        ds = divisorsRT (p - 1)
        n = p'
        -- Build one solvable group per divisor of p-1
        mkAffine d =
            let trans = fromMapping [fromIntegral ((i + 1) `mod` p) | i <- [0 .. p - 1]]
                scaleFactor = modExpRT g ((p - 1) `div` d) p
                scale = fromMapping [fromIntegral ((scaleFactor * i) `mod` p) | i <- [0 .. p - 1]]
                gens = if d == 1 then [trans] else [trans, scale]
                gName
                    | d == 1     = "Z" ++ show p
                    | d == 2     = "D" ++ show p
                    | d == p - 1 = "AGL(1," ++ show p ++ ")"
                    | otherwise  = "Z" ++ show p ++ ":Z" ++ show d
                cFactors = map fromIntegral (primeFactorsWithMultRT d ++ [p])
             in TransitiveGroup
                    { tgName = gName
                    , tgDegree = n
                    , tgOrder = p * d
                    , tgGenerators = gens
                    , tgSolvable = True
                    , tgMaximalSupergroups = []
                    , tgCompositionFactors = cFactors
                    }
        solvableGroups = map mkAffine ds
        -- A_p
        ap = TransitiveGroup
            { tgName = "A" ++ show p
            , tgDegree = n
            , tgOrder = product [1 .. p] `div` 2
            , tgGenerators =
                [ fromCycles n [[0 .. n - 1]]
                , fromCycles n [[0, 1, 2]]
                ]
            , tgSolvable = p' < 5
            , tgMaximalSupergroups = []
            , tgCompositionFactors = []
            }
        -- S_p
        sp = TransitiveGroup
            { tgName = "S" ++ show p
            , tgDegree = n
            , tgOrder = product [1 .. p]
            , tgGenerators =
                [ fromCycles n [[0 .. n - 1]]
                , fromCycles n [[0, 1]]
                ]
            , tgSolvable = p' < 4
            , tgMaximalSupergroups = []
            , tgCompositionFactors = []
            }
        allGroups = sort $ solvableGroups ++ [ap, sp]
        -- Assign maximal supergroups
        indexed = zip [0 :: Int ..] allGroups
        assignSuper (myIdx, tg) =
            let myOrd = tgOrder tg
                cands = [(i, tgOrder cg) | (i, cg) <- indexed, i /= myIdx, tgOrder cg > myOrd, tgOrder cg `mod` myOrd == 0]
                isMax (_, superOrd) =
                    not $ any (\(_, midOrd) -> midOrd > myOrd && midOrd < superOrd && superOrd `mod` midOrd == 0 && midOrd `mod` myOrd == 0) cands
             in tg{tgMaximalSupergroups = [i | (i, _) <- cands, isMax (i, tgOrder (snd (indexed !! i)))]}
     in map assignSuper indexed

instance Ord TransitiveGroup where
    compare a b = compare (tgOrder a) (tgOrder b)

------------------------------------------------------------------------
-- Composition series
------------------------------------------------------------------------

{- | For a solvable transitive group, return the composition series as a
list of subgroups (each given by its generating set), descending from
\(G\) to the trivial group \(\{1\}\):

\[
  G = G_0 \supset G_1 \supset \cdots \supset G_k = \{1\}
\]

Each consecutive quotient \(G_i / G_{i+1}\) is cyclic of prime order
(matching the entries of 'tgCompositionFactors').  This chain drives
radical tower descent: to express the roots in radicals, adjoin a
\(|G_i / G_{i+1}|\)-th root at each step, working from the bottom of
the chain upward.

Returns 'Nothing' for non-solvable groups.

For degree-5 groups, uses the hard-coded series. For prime-degree groups,
computes the series from the affine structure.
-}
compositionSeries :: TransitiveGroup -> Maybe [[Perm]]
compositionSeries tg
    | not (tgSolvable tg) = Nothing
    | otherwise = case tgName tg of
        -- Fast path: degree-5 hard-coded
        "C5" -> Just [tgGenerators tg, []]
        "D5" ->
            Just
                [ tgGenerators tg
                , [fromCycles 5 [[0, 1, 2, 3, 4]]] -- C5
                , []
                ]
        "F20" ->
            Just
                [ tgGenerators tg
                , -- D5 inside F20
                    [ fromCycles 5 [[0, 1, 2, 3, 4]]
                    , fromCycles 5 [[1, 4], [2, 3]]
                    ]
                , [fromCycles 5 [[0, 1, 2, 3, 4]]] -- C5
                , []
                ]
        -- General prime-degree affine groups
        _ | isPrime (fromIntegral (tgDegree tg)) ->
                compositionSeriesPrime tg
          | otherwise -> Nothing

{- | Composition series for a solvable affine subgroup of \(S_p\).

The group has the form \(\mathbb{Z}/p \rtimes H\) where \(H\) is cyclic
of order \(d\). The series descends through subgroups of \(H\) by
removing one prime factor at a time, then drops to \(\{1\}\).
-}
compositionSeriesPrime :: TransitiveGroup -> Maybe [[Perm]]
compositionSeriesPrime tg =
    let p = fromIntegral (tgDegree tg) :: Integer
        n = tgDegree tg
        g = primitiveRootRT p
        d = tgOrder tg `div` p
        dFactors = primeFactorsWithMultRT d
        -- Chain of divisors: d, d/q1, d/(q1*q2), ..., 1
        dChain = scanl (\acc q -> acc `div` q) d dFactors
        trans = fromCycles n [[0 .. n - 1]]
        mkGens d'
            | d' <= 1 = [trans]
            | otherwise =
                let sf = modExpRT g ((p - 1) `div` d') p
                    scale = fromMapping [fromIntegral ((sf * i) `mod` p) | i <- [0 .. p - 1]]
                 in [trans, scale]
     in Just (map mkGens dChain ++ [[]])

File: ocaml/lib/transitive_group.ml
(** Transitive subgroups of S_n for Galois group identification.

    For prime p, solvable transitive subgroups of S_p are exactly the
    affine groups Z/p ⋊ H, one per divisor d of p-1, where |H| = d.
    These are computed at runtime from the AGL(1,p) structure.

    The lattice also includes A_p and S_p (non-solvable for p >= 5). *)

type t = {
  name : string;
  degree : int;
  order : int;
  generators : Permutation.t list;
  solvable : bool;
  maximal_supergroups : int list;
  composition_factors : int list;
}

(** {2 Helpers} *)

let is_prime n =
  if n < 2 then false
  else if n < 4 then true
  else
    let rec check d =
      if d * d > n then true
      else if n mod d = 0 then false
      else check (d + if d = 2 then 1 else 2)
    in
    check 2

let factorial n =
  let rec go acc = function
    | 0 | 1 -> acc
    | k -> go (acc * k) (k - 1)
  in
  go 1 n

(** Modular exponentiation: base^exp mod m. *)
let mod_exp base exp m =
  let rec go result b e =
    if e = 0 then result
    else if e mod 2 = 1 then go ((result * b) mod m) b (e - 1)
    else go result ((b * b) mod m) (e / 2)
  in
  go 1 (base mod m) exp

(** Sorted positive divisors of n. *)
let divisors n =
  let result = ref [] in
  let d = ref 1 in
  while !d * !d <= n do
    if n mod !d = 0 then begin
      result := !d :: !result;
      if !d * !d <> n then result := (n / !d) :: !result
    end;
    incr d
  done;
  List.sort Int.compare !result

(** Primitive root modulo a prime p. *)
let primitive_root p =
  let phi = p - 1 in
  let factors =
    let pos = Positive.of_int_exn phi in
    List.map fst (Prime_factors.factorise pos)
  in
  let is_prim_root g =
    List.for_all (fun q -> mod_exp g (phi / q) p <> 1) factors
  in
  let rec find g =
    if g >= p then failwith "primitive_root: none found"
    else if is_prim_root g then g
    else find (g + 1)
  in
  find 2

(** Prime factorisation as a flat list with multiplicity. *)
let prime_factors_with_mult n =
  if n <= 1 then []
  else
    let pos = Positive.of_int_exn n in
    List.concat_map (fun (p, e) ->
      List.init e (fun _ -> p)
    ) (Prime_factors.factorise pos)

(** {2 Runtime prime group computation} *)

(** All transitive subgroups of S_p for a prime p.

    The solvable subgroups are Z/p ⋊ H for each divisor d of p-1,
    plus A_p and S_p (non-solvable for p >= 5). *)
let trans_groups_of_prime p =
  let g = primitive_root p in
  let ds = divisors (p - 1) in
  let n = p in
  (* Build one solvable group per divisor of p-1 *)
  let mk_affine d =
    let trans = Permutation.from_mapping
      (List.init p (fun i -> (i + 1) mod p)) in
    let scale_factor = mod_exp g ((p - 1) / d) p in
    let scale = Permutation.from_mapping
      (List.init p (fun i -> (scale_factor * i) mod p)) in
    let gens = if d = 1 then [trans] else [trans; scale] in
    let name =
      if d = 1 then Printf.sprintf "Z%d" p
      else if d = 2 then Printf.sprintf "D%d" p
      else if d = p - 1 then Printf.sprintf "AGL(1,%d)" p
      else Printf.sprintf "Z%d:Z%d" p d
    in
    let c_factors = prime_factors_with_mult d @ [p] in
    { name;
      degree = n;
      order = p * d;
      generators = gens;
      solvable = true;
      maximal_supergroups = [];  (* assigned later *)
      composition_factors = c_factors }
  in
  let solvable_groups = List.map mk_affine ds in
  (* A_p *)
  let ap = {
    name = Printf.sprintf "A%d" p;
    degree = n;
    order = factorial p / 2;
    generators = [
      Permutation.from_cycles n [List.init n Fun.id];
      Permutation.from_cycles n [[0; 1; 2]];
    ];
    solvable = p < 5;
    maximal_supergroups = [];
    composition_factors = [];
  } in
  (* S_p *)
  let sp = {
    name = Printf.sprintf "S%d" p;
    degree = n;
    order = factorial p;
    generators = [
      Permutation.from_cycles n [List.init n Fun.id];
      Permutation.from_cycles n [[0; 1]];
    ];
    solvable = p < 4;
    maximal_supergroups = [];
    composition_factors = [];
  } in
  let all_groups =
    List.sort (fun a b -> Int.compare a.order b.order) (solvable_groups @ [ap; sp])
  in
  (* Assign maximal supergroups *)
  let indexed = List.mapi (fun i g -> (i, g)) all_groups in
  List.map (fun (my_idx, tg) ->
    let my_ord = tg.order in
    let cands =
      List.filter_map (fun (i, cg) ->
        if i <> my_idx && cg.order > my_ord && cg.order mod my_ord = 0 then
          Some (i, cg.order)
        else None
      ) indexed
    in
    let is_max (_, super_ord) =
      not (List.exists (fun (_, mid_ord) ->
        mid_ord > my_ord && mid_ord < super_ord &&
        super_ord mod mid_ord = 0 && mid_ord mod my_ord = 0
      ) cands)
    in
    let maximal = List.filter_map (fun (i, o) ->
      if is_max (i, o) then Some i else None
    ) cands in
    { tg with maximal_supergroups = maximal }
  ) indexed

(** {2 Public API} *)

(** All transitive subgroups of S_n (up to conjugacy) for degree n,
    sorted by increasing group order.

    For prime n, uses runtime AGL(1,n) computation.
    Returns the empty list for unsupported composite degrees. *)
let trans_groups_of_degree n =
  if n >= 3 && is_prime n then trans_groups_of_prime n
  else []

(** Find the transitive group(s) of the given degree and order. *)
let trans_group_by_order deg ord =
  List.filter (fun g -> g.order = ord) (trans_groups_of_degree deg)

(** Test whether a transitive group is solvable. *)
let is_solvable tg = tg.solvable

(** {2 Composition series} *)

(** For a solvable affine subgroup of S_p, return the composition series
    as a list of generating sets descending from G to {1}.

    The group has the form Z/p ⋊ H where H is cyclic of order d.
    The series descends through subgroups of H by removing one prime
    factor at a time. *)
let composition_series_prime tg =
  let p = tg.degree in
  let n = p in
  let g = primitive_root p in
  let d = tg.order / p in
  let d_factors = prime_factors_with_mult d in
  (* Chain of divisors: d, d/q1, d/(q1*q2), ..., 1 *)
  let d_chain =
    List.fold_left (fun acc q ->
      let last = List.hd acc in
      (last / q) :: acc
    ) [d] d_factors |> List.rev
  in
  let trans = Permutation.from_cycles n [List.init n Fun.id] in
  let mk_gens d' =
    if d' <= 1 then [trans]
    else
      let sf = mod_exp g ((p - 1) / d') p in
      let scale = Permutation.from_mapping
        (List.init p (fun i -> (sf * i) mod p)) in
      [trans; scale]
  in
  List.map mk_gens d_chain @ [[]]

(** Return the composition series for a solvable transitive group.
    Returns [None] for non-solvable groups. *)
let composition_series tg =
  if not tg.solvable then None
  else if is_prime tg.degree then
    Some (composition_series_prime tg)
  else None

(** Symmetric group S_n. *)
let symmetric n =
  let transposition i j =
    Array.init n (fun k ->
      if k = i then j
      else if k = j then i
      else k)
  in
  let gens = if n <= 1 then []
    else [
      transposition 0 1;
      Array.init n (fun i -> (i + 1) mod n)
    ] in
  { name = Printf.sprintf "S%d" n;
    degree = n;
    generators = gens;
    order = factorial n;
    solvable = n < 4;
    maximal_supergroups = [];
    composition_factors = [] }
